### PR TITLE
Implement variance RFC #738 

### LINKED ISSUE #22212: Tracking issue for Support variance for type parameters (RFC 738)
https://github.com/rust-lang/rfcs/pull/738

### PR DESCRIPTION
**Pull request by nikomatsakis**
This PR:
- converts the markers to the new `PhantomData`/`PhantomFn` scheme;
- corrects various errors in variance inference, which was not always taking the correct things into account;
- integrates results variance into trait matching, coherence, subtyping.

This is a [breaking-change]:
- The old markers are removed in favor of `PhantomData`. It is possible to simulate all of the old markers:
  - `ContravariantLifetime<'a>` -> `PhantomData<&'a ()>` (or just `&'a ()`, actually)
  - `InvariantLifetime<'a>` -> `PhantomData<Cell<&'a ()>>`
  - `CovariantType<T>` -> `PhantomData<T>`
  - `ContravariantType<T>` -> `PhantomData<fn(T)>`
  - `InvariantType<T>` -> `PhantomData<Cell<T>>`
- It is now illegal to have unused type and lifetime parameters. Typically the best fix is either to remove the parameter or to add an instance of `PhantomData` to your struct (or `PhantomFn` / `MarkerTrait` to your trait). See rust-lang/rfcs#738 for more background.
- Technically, variance can cause some impls that used to be legal to be illegal due to subtyping. This is highly unlikely to affect you.

Fixes #22212.

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Grr, something in the last rebase started causing some (minor looking) test failures. I'll track those down.

File: src/libarena/lib.rs [hunk 1]
```diff
@@ -96,7 +96,7 @@ pub struct Arena<'longer_than_self> {
     head: RefCell<Chunk>,
     copy_head: RefCell<Chunk>,
     chunks: RefCell<Vec<Chunk>>,
-    _invariant: marker::InvariantLifetime<'longer_than_self>,
+    _marker: marker::PhantomData<*mut &'longer_than_self()>,
```
Comment (pkondzior) [inline src/libarena/lib.rs]: Space after lifetime name is missing ?

**Comment by nikomatsakis** [comment]
Rebased. Make check passes locally now.

**Comment by frewsxcv** [comment]
Needs a rebase

File: src/libcore/marker.rs [hunk 2]
```diff
@@ -424,3 +222,80 @@ pub struct NoCopy;
...
+
+/// `PhantomData` is a way to tell the compiler about fake fields.
+/// The idea is that if the compiler encounters a `PhantomData<T>`
+/// instance, it will behave *as if* an instance of the type `T` were
+/// present for the purpose of various automatic analyses.
+///
+/// FIXME. Better documentation needed here!
+#[cfg(not(stage0))]
+#[lang="phantom_data"]
+pub struct PhantomData<T:?Sized>;
```
Comment (alexcrichton) [inline src/libcore/marker.rs]: Could you go ahead an mark this `#[stable]` as well? I think it'd been debated enough that it's probably earned it. (you may want to leave `PhantomFn` explicitly `#[unstable]` for now though).

File: src/libcore/marker.rs [hunk 3]
```diff
@@ -198,216 +202,10 @@ pub trait Copy {
...
-
-/// As `InvariantType`, but for lifetime parameters. Using
-/// `InvariantLifetime<'a>` indicates that it is not ok to
-/// substitute any other lifetime for `'a` besides its original
-/// value. This is appropriate for cases where you have an unsafe
-/// pointer that is actually a pointer into memory with lifetime `'a`,
-/// and this pointer is itself stored in an inherently mutable
-/// location (such as a `Cell`).
-#[unstable(feature = "core",
-           reason = "likely to change with new variance strategy")]
```
Comment (alexcrichton) [inline src/libcore/marker.rs]: Could these be `#[deprecated]` instead of removed if possible? (with messages pointing out what to use)

**Comment by pnkfelix** [comment]
@nikomatsakis okay i'm done with the review.  Some nits and minor questions, but I would not object to you telling bors that this PR as-is is `r=pnkfelix`.

**Comment by alexcrichton** [comment]
This was merged in #22541

**Pull request closed without merging**